File: core/src/rules/validate-property-value-type.ts
// core/src/rules/validate-property-value-type.ts
//
// Atomic rule: checks whether a property's value matches its expected
// schema.org value-types (per the registry's `valueTypes` array).
//
// Cycle 1 model — approximate, sufficient for Product Layer 1:
//   - JS string   → matches Text, URL, Date, DateTime, Time, or any
//                   non-primitive (object) type (schema.org allows
//                   object-typed properties to be referenced by URL).
//   - JS number   → matches Number or Integer.
//   - JS boolean  → matches Boolean.
//   - JS object   → matches any non-primitive type (deeper @type
//                   verification deferred to later cycle).
//   - JS array    → not supported in cycle 1; emits the mismatch.
//   - null        → returns empty (orchestrator decides whether
//                   missing-property is a separate concern).
//
// Returns Issue[] (possibly empty). Pure function.

import type { Issue } from "../types.js";
import { invalidPropertyValueType } from "../errors.js";

const SCHEMA_PRIMITIVES = new Set([
  "Text",
  "URL",
  "Number",
  "Integer",
  "Boolean",
  "Date",
  "DateTime",
  "Time",
]);

function isObjectType(typeName: string): boolean {
  return !SCHEMA_PRIMITIVES.has(typeName);
}

const STRING_PRIMITIVES = new Set(["Text", "URL", "Date", "DateTime", "Time"]);

export function validatePropertyValueType(
  typeName: string,
  propertyName: string,
  value: unknown,
  valueTypes: string[],
): Issue[] {
  if (value === null) {
    return [];
  }

  const fail = (): Issue[] => [
    invalidPropertyValueType(typeName, propertyName, valueTypes, value),
  ];

  if (typeof value === "string") {
    // String accepted by any string primitive or any non-primitive (URL ref).
    const ok = valueTypes.some(
      (t) => STRING_PRIMITIVES.has(t) || isObjectType(t),
    );
    return ok ? [] : fail();
  }

  if (typeof value === "number") {
    const ok = valueTypes.some((t) => t === "Number" || t === "Integer");
    return ok ? [] : fail();
  }

  if (typeof value === "boolean") {
    const ok = valueTypes.some((t) => t === "Boolean");
    return ok ? [] : fail();
  }

  if (typeof value === "object" && !Array.isArray(value)) {
    const ok = valueTypes.some((t) => isObjectType(t));
    return ok ? [] : fail();
  }

  // Arrays, functions, symbols, bigints — out of scope for cycle 1.
  return fail();
}

File: python/src/schema_audit/rules/validate_property_value_type.py
"""Atomic rule: check a property's value against its expected
schema.org value-types.

Mirrors ``core/src/rules/validate-property-value-type.ts`` 1:1.

Cycle 9 model (approximate, sufficient for the curated coverage):

- Python ``str``     → matches ``Text``, ``URL``, ``Date``, ``DateTime``,
  ``Time``, or any non-primitive (object) type — schema.org allows
  object-typed properties to be referenced by URL.
- Python ``int``/``float`` (but NOT ``bool``) → matches ``Number`` or
  ``Integer``.
- Python ``bool``    → matches ``Boolean``.
- Python ``dict``    → matches any non-primitive type (deeper ``@type``
  verification deferred).
- Python ``list``    → not supported in cycle 9; emits the mismatch.
- ``None``           → returns empty (the orchestrator decides whether a
  missing property is a separate concern).
"""

from __future__ import annotations

from ..errors import invalid_property_value_type
from ..types import Issue

_SCHEMA_PRIMITIVES = frozenset(
    {"Text", "URL", "Number", "Integer", "Boolean", "Date", "DateTime", "Time"}
)
_STRING_PRIMITIVES = frozenset({"Text", "URL", "Date", "DateTime", "Time"})


def _is_object_type(type_name: str) -> bool:
    return type_name not in _SCHEMA_PRIMITIVES


def validate_property_value_type(
    type_name: str,
    property_name: str,
    value: object,
    value_types: list[str],
) -> list[Issue]:
    if value is None:
        return []

    def fail() -> list[Issue]:
        return [invalid_property_value_type(type_name, property_name, value_types, value)]

    # `bool` is a subclass of `int` in Python — check it first.
    if isinstance(value, bool):
        ok = any(t == "Boolean" for t in value_types)
        return [] if ok else fail()

    if isinstance(value, str):
        ok = any(t in _STRING_PRIMITIVES or _is_object_type(t) for t in value_types)
        return [] if ok else fail()

    if isinstance(value, int | float):
        ok = any(t in ("Number", "Integer") for t in value_types)
        return [] if ok else fail()

    if isinstance(value, dict):
        ok = any(_is_object_type(t) for t in value_types)
        return [] if ok else fail()

    # Lists, anything else — out of scope (orchestrator may handle list-fan-out).
    return fail()
